Source/WebCore/html/CanvasBase.cpp

```diff
@@ -199,6 +200,12 @@ void CanvasBase::notifyObserversCanvasChanged(const std::optional<FloatRect>& re
         observer.canvasChanged(*this, rect);
 }
 
+void CanvasBase::didDraw(const std::optional<FloatRect>& rect, const ShouldApplyPostProcessingToDirtyRect shouldApplyPostProcessingToDirtyRect)
+{
+    if (shouldInjectNoiseBeforeReadback() && shouldApplyPostProcessingToDirtyRect == ShouldApplyPostProcessingToDirtyRect::Yes && rect && !m_postProcessDirtyRect.contains(*rect))
```

Comment:
`if (!a.contains(b)) a.unite(b);` == `a.unite(b);`
I.e. no need to check `contains`, at that is equal to just doing the `unite` in terms of complexity

I think it would be better to categorize the data in rect as "something needing postprocess" and "something does not need post process". 

So we have a goal to give out prestine putImageData back in some cases. 

With your patch, consider:
 empty canvas + full image putImageData -> you can get prestine putImageData back
 empty canvas + draw line + full image putimageData -> you cannot get prestine putImageData back

It is not logical. If putImageData is "not need postprocess", you should always get prestine putImageData back in relevant, same cases.

Reply:
> if (!a.contains(b)) a.unite(b); == a.unite(b);

Good point.

> So we have a goal to give out prestine putImageData back in some cases.

I was approaching this as best-effort because I made a wrong assumption that `alpha != 1` would cause compositing. That is obviously wrong in retrospect.

> If putImageData is "not need postprocess", you should always get prestine putImageData back in relevant, same cases.

I agree this is the ideal and expected behavior, but I'll add this in a follow-up patch. Supporting this makes the implementation more complicated.